packages/flutter_test/lib/src/tree_traversal.dart

```diff
@@ -0,0 +1,165 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/semantics.dart';
+import 'package:flutter/widgets.dart';
+
+/// Provides an iterable that efficiently returns all the [Element]s
+/// rooted at the given [Element]. See [CachingIterable] for details.
+///
+/// This function must be called again if the tree changes. You cannot
+/// call this function once, then reuse the iterable after having
+/// changed the state of the tree, because the iterable returned by
+/// this function caches the results and only walks the tree once.
+///
+/// The same applies to any iterable obtained indirectly through this
+/// one, for example the results of calling `where` on this iterable
+/// are also cached.
+Iterable<Element> collectAllElementsFrom(
+  Element rootElement, {
+  required bool skipOffstage,
+}) {
+  return CachingIterable<Element>(_DepthFirstElementTreeIterator(rootElement, !skipOffstage));
+}
+
+/// Provides an iterable that efficiently returns all the [SemanticsNode]s
+/// rooted at the given [SemanticsNode]. See [CachingIterable] for details.
+///
+/// By default, this will traverse the semantics tree in semantic traversal
+/// order, but the traversal order can be changed by passing in a different
+/// value to [order].
+///
+/// If [includeRoot] is false, the root of the semantics tree will not be
+/// included. This value is true by default and will include the root.
```

Comment:
For my own understanding: why do all the semantics-related methods have a way to exclude the root?

Reply:
It's related to the `matchRoot` arguments for ancestor/descendant finders.

The similar element traversal just doesn't let you include the root that you're descending from in the search, so when `matchRoot` is true, you have to traverse to gather descendants, then add in the roots to the candidates list.

With the `includeRoot` argument on `collectAllSemanticsNodesFrom`, we can get the right candidates back from the traversal without having to add anything later.